# Ejercicio 10

> Por: Mapachana

**Toma tu número p de la lista publicada para este ejercicio.**

In [1]:
p = 77770081
p

77770081

In [2]:
import math

**i) Calcula el símbolo de Jacobi $\left( \frac{-11}{p} \right)$. Si sale 1, usa el algoritmo de Tonelli-Shanks para hallar soluciones a la congruencia $x^2 \equiv -11 \mod p$.**

In [3]:
jacobi_symbol(-11, p)

1

Tenemos que $\left( \frac{-11}{p}\right)=1$ luego aplicamos el algoritmo de Tonelli-Shanks para hallar las soluciones de $x^2\equiv -11\mod p$.

Para ello, comenzamos factorizando $p-1=2^5\cdot 3^2\cdot 5 \cdot 53\cdot 1019$

Para ello vamos a usar el método $\rho$ de Polard. Comenzamos sacando los factores 2, y aplicamos el algoritmo a 2430315. Para factorizar este número han sido necesarias un total de 11 iteraciones.

Finalmente, aplicamos Lucas-Lehmer, obtenemos que 17 es elemento primitivo para $p$, ya que $17^{p-1}\equiv 1\mod p$ y $17^{\frac{p-1}{p}}\not\equiv 1\mod p $ para $p\in\{2,3,5, 53, 1019\}$ pues::

\begin{verbatim}
17^(p-1)/ 2 = 77770080 mod p
17^(p-1)/ 3 = 58134188 mod p
17^(p-1)/ 5 = 66432901 mod p
17^(p-1)/ 53 = 68065795 mod p
17^(p-1)/ 1019 = 65224721 mod p
\end{verbatim}

Luego $p$ es primo.

Como $p\equiv 1\mod 8$ tenemos que usar Tonelli-Shanks.

Hemos visto además que $p-1=2^5\cdot 2430315$, y  como $(-11)^{2430315} \equiv 158982 \mod p \not\equiv 1\mod p$.

In [4]:
(p-1).factor()

2^5 * 3^2 * 5 * 53 * 1019

In [5]:
p.mod(8)

1

In [6]:
pow(-11, 2430315, p)

158982

In [16]:
def orden(t, p):
    for i in range(1, p):
        if pow(t,i,p) == 1:
            return i
        
def exp_orden(n, p):
    acu = 0
    
    while n != 1:
        n = n//p
        acu += 1
        
    return acu

In [8]:
p = 77770081
print(orden(158982, p))
print(orden(4653122, 27213937))

4
4


In [9]:
print(p)
print(jacobi_symbol(1, p))
print(jacobi_symbol(2, p))
print(jacobi_symbol(3, p))
print(jacobi_symbol(4, p))
print(jacobi_symbol(5, p))
print(jacobi_symbol(6, p))
print(jacobi_symbol(7, p))
print(jacobi_symbol(8, p))
print(jacobi_symbol(9, p))
print(jacobi_symbol(10, p))
print(jacobi_symbol(11, p))
print(jacobi_symbol(12, p))
print(jacobi_symbol(13, p))
print(jacobi_symbol(14, p))
print(jacobi_symbol(15, p))
print(jacobi_symbol(16, p))
print(jacobi_symbol(17, p))
print(jacobi_symbol(18, p))
print(jacobi_symbol(19, p))
print(jacobi_symbol(20, p))

77770081
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
-1
1
-1
1


In [10]:
n = 1
while jacobi_symbol(n, p) != -1:
    n = n + 1 
print(n)

17


Y este tiene orden $2^2 \mod p$.

Calculamos ahora un no residuo cuadrático módulo p, y el primero es $n=17$.

Por tanto un generador del 2-subgrupo de Sylow $G=\mathbb{Z}_{2^{17}}$.

In [23]:
p = 77770081
q = 2430315
e = 5
gen = 17

#p = 77646353
#q = 4852897
#e = 4
#gen = 3

z = pow(gen, q, p)
t = pow(-11, q, p)
i = exp_orden(int(orden(t, p)), 2)
r = pow(-11, (q+1)//2, p)

print("z: {}, t: {}, i: {}, r: {}".format(z,t,i,r))

aux = pow(2, e-i-1)
b = pow(z, aux, p)
t1 = (t*pow(b,2)).mod(p)
i1 = exp_orden(orden(t1, p), 2)
r1 = (r*b).mod(p)

print("b: {}, t1: {}, i1: {}, r1: {}".format(b,t1,i1,r1))

print("Soluciones")
print(r1, p-r1)


print(p-11)
print(pow(r1, 2, p))
print(pow(p-r1, 2, p))

z: 55328379, t: 158982, i: 2, r: 29971170
b: 12935680, t1: 1, i1: 0, r1: 796425
Soluciones
796425 76973656
77770070
77770070
77770070


In [22]:
def calcular_z(p):
    n = 1
    while jacobi_symbol(n, p) != -1:
        n = n + 1 
    
    e = factor(p-1)[0][1]
    q = (p-1)//2^e
    z = (n^q).mod(p)
    
    return z
    

def tonelli_shanks(a, p):
    e = factor(p-1)[0][1]
    print("e = " + str(e))
    q = (p-1)//2^e
    print("q = " + str(q))
    t = a^q
    print("t mod p = " + str(t.mod(p)))
    z = calcular_z(p)
    print("z mod p = " + str(z))
    r = a.powermod((q+1)//2, p)
    print("r mod p = " + str(r))
    r_1 = r
    
    while t.mod(p) != 1:
        print("\nInicio de bucle:")
        i = factor(orden(t, p))[0][1]
        print("\tOrden de t = " + str(orden(t, p)) + " => " + "i = " + str(i))
        b = z.powermod(2^(e - i - 1), p)
        print("\tb mod p = " + str(b))
        b_cuadrado = b.powermod(2, p)
        print("\tb^2 mod p = " + str(b_cuadrado))
        r_1 = (r_1 * b).mod(p)
        print("\tr_1 mod p = " + str(r_1))
        t_1 = (t * b_cuadrado).mod(p)
        print("\tt_1 mod p = " + str(t_1))

        t = t_1
    
    x_1 = r_1.mod(p)
    x_2 = (p - x_1).mod(p)
    
    return [x_1, x_2]

soluciones = tonelli_shanks(-11, p)
soluciones


e = 5
q = 2430315
t mod p = 158982
z mod p = 55328379
r mod p = 29971170

Inicio de bucle:
	Orden de t = 4 => i = 2
	b mod p = 12935680
	b^2 mod p = 77611099
	r_1 mod p = 796425
	t_1 mod p = 1


[796425, 76973656]

Luego la soluciones son $796425$ y $76973656$. ($r_1$ y $p-r_1$

**ii) Usa una de esas soluciones para factorizar el ideal principal, (p) =(p,n+$\sqrt{-11}$)(p,n+$\sqrt{-11}$) como producto de dos ideales.**

Tomamos la solución impar $n=796425$ y como $p$ es un impar primo que no divide a -11 entonces $(p)=\left(p, 796425+\sqrt{-11}\right)\left(796425-\sqrt{-11}\right)$

In [27]:
n = 796425

**iii) Aplica el algoritmo de Cornachia- Smith modificado a 2p y n para encontrar una solucíón a la ecuación diofántica $4p =x^2 + 11y^2$ y la usas para encontrar una factorización de p en a.e. del cuerpo $\mathbb{Q}(\sqrt{-11})$.**

In [32]:
def cornachia_smith(p, n):
    siguiente_Div = 2*p
    siguiente_div = n
    
    resto = 2*p
    i = 1
    
    while not resto < 2*sqrt(p):
        cociente, resto = divmod(siguiente_Div, siguiente_div)
        print("Paso " + str(i) + ":   " + str(siguiente_Div) + " = " + str(cociente) + "*" + str(siguiente_div) + " + " + str(resto))

        siguiente_Div = siguiente_div
        siguiente_div = resto
        i = i + 1
        
    return resto    
    
# ejemplo diapos
#cornachia_smith(27213937, 21148483)
# ejemplo wuolah
#print()
#cornachia_smith(77646353, 34279617)
print(2*math.sqrt(p))

17637.46931960478


Aplicamos el algoritmo de Cornachia-Smith:

In [31]:
cornachia_smith(p,n)

Paso 1:   155540162 = 195*796425 + 237287
Paso 2:   796425 = 3*237287 + 84564
Paso 3:   237287 = 2*84564 + 68159
Paso 4:   84564 = 1*68159 + 16405


16405

Tras 4 divisiones obtenemos el resto $x=16405$. Por tanto, podemos hallar $y$ despejando de la ecuación:
$$4p =x^2 + 11y^2$  \LeftRightarrow y = \sqrt{\frac{4p-x^2}{11}}=1953$

Así, se cumple  $4p = 16405^2+11\cdot 1953^2$.

x = 16405
math.sqrt((4*p-pow(x,2))/11)

In [35]:
print(pow(16405, 2)+11*pow(1953, 2))
print(4*p)

311080324
311080324


Luego tenemos que la factorización de $p$ en a.e. de $\mathbb{Q}[\sqrt{p}]$ es
$$p= \left( \frac{x+y\sqrt{-11}}{2}\right) \left( \frac{x-y\sqrt{-11}}{2}\right) = \left( \frac{16405+1953\sqrt{-11}}{2}\right) \left( \frac{16405-1953\sqrt{-11}}{2}\right)$$

**iv) ¿ Son principales tus ideales (p,n+$\sqrt{-11}$)(p,n+$\sqrt{-11}$)**

Son principales ya que $\mathbb{q}[\aqrt{-11}]$ es DIP.